mtags/src/main/scala-2/scala/meta/internal/pc/PcDefinitionProvider.scala

```diff
@@ -131,4 +131,58 @@ class PcDefinitionProvider(val compiler: MetalsGlobal, params: OffsetParams) {
         loop(tree)
     }
   }
+
+  /** A fully attributed tree located at position `pos` */
+  private def typedTreeAt(pos: Position, unit: RichCompilationUnit): Tree = {
+    typeCheck(unit)
+    new Locator(pos).locateIn(unit.body)
+  }
+
+  /**
+   * Taken from https://github.com/scala/scala/blob/8a2cf63ee5bad8c8c054f76464de0e10226516a0/src/reflect/scala/reflect/internal/Positions.scala#L332
+   *  with modifications.
+   *  Added additional `isMoreNarrow` check.
+   */
+  class Locator(pos: Position) extends Traverser {
+    var last: Tree = _
+    def locateIn(root: Tree): Tree = {
+      this.last = EmptyTree
+      traverse(root)
+      this.last
+    }
+    protected def isEligible(t: Tree): Boolean = !t.pos.isTransparent
+
+    private def isMoreNarrow(pos: Position): Boolean = {
+      last match {
+        case EmptyTree => true
+        case t => t.pos.includes(pos)
+      }
+    }
+
+    override def traverse(t: Tree) {
+      t match {
+        case tt: TypeTree
+            if tt.original != null && (tt.pos includes tt.original.pos) =>
+          traverse(tt.original)
+        case _ =>
+          if (t.pos includes pos) {
+            if (isEligible(t) && isMoreNarrow(t.pos))
```

Comment:
The difference with original `typedTreeAt` and `Locator` is in this line. I added `isMoreNarrow` check.

##### Explanation

Having the following source
```scala
class Main {
  val a = 42
  val all = 42 :: List(1, 2)
}
```
The left side tree for `val all` is:
```
Block(
  List(
    ValDef(.. x$1, ... , Select(This(Main), a)) // 1 <- synthetic 
    Apply(...) // 2 call `::` in reversed order using synthetic - `List(1,2) :: x$1`
  )
)
```

The problem that is in scala2.12  `Locator` first receives 1(synthetic) and then 2(call `::`).
Both of them includes definition pos so 2(call `::`) tree overwrites the previous one that leads to having wrongly located tree.